backend/src/routes/api/scene.js

```diff
@@ -27,10 +27,45 @@ router.get("/full/:sceneId", async (req, res) => {
 // Retrieve all scenes of a scenario
 router.get("/", async (req, res) => {
   const scenes = await retrieveSceneList(req.params.scenarioId);
-
   res.json(scenes);
 });
 
+// Get graph
+router.get("/graph", async (req, res) => {
+  const scenes = await retrieveSceneList(req.params.scenarioId);
+  let graph = {};
```

Comment:
You should separate this out into its own class, with the responsibility split as following:

the routers file:
- takes in information from the request
- retrieves things from the database

the new "model" file:
- takes a set of scene's in constructor, as well as the root node
- in constructor, forms underlying graph data structure
- has methods for getting stats (getCurrentProgress(currentSceneId): float)

As an interface for the new model file, maybe call it ScenarioGraph:

```
ScenarioGraph.js

class ScenarioGraph {
  
  rootNode: Node;
  
  constructor(rootScene: Scene, otherScenes: Scene[]) {
    // Build node
  }

  getDistanceBetween(startScene sceneId, endScene sceneId);
}